[DAGCombiner] Improve FMin/FMax DAGCombines

LewisCrawford opened the pull request:
Add several improvements to DAGCombine patterns for fmin/fmax:
 * Fix incorrect results due to minimumnum not being marked as IsMin
    - e.g. nnan minimumnum(x, +inf) returned +inf instead of x
 * Fix incorrect results checking maximumnum for vecreduce patterns
 * Make maxnum/minnum return QNaN if one input is SNaN instead of X
 * Quiet SNaN inputs when propagating them e.g. maximum(x, SNaN) = QNaN
 * Update comments to mark when SNaN propagation is being ignored

LewisCrawford committed:
[DAGCombiner] Improve FMin/FMax DAGCombines

Add several improvements to DAGCombine patterns for fmin/fmax:
 * Fix incorrect results due to minimumnum not being marked as IsMin
    - e.g. nnan minimumnum(x, +inf) returned +inf instead of x
 * Fix incorrect results checking maximumnum for vecreduce patterns
 * Make maxnum/minnum return QNaN if one input is SNaN instead of X
 * Quiet SNaN inputs when propagating them e.g. maximum(x, SNaN) = QNaN
 * Update comments to mark when SNaN propagation is being ignored

Artem-B reviewed (commented):
LGTM overal, with a test nit.

LewisCrawford committed:
Ensure test functions are not NOPs

Previously, when optimizing fmax(x, nan) -> x, the
functions would return immediately, since the input
parameter x to the function is already in the necessary
register for output.

To make the transformations more explicit, instead add a 2nd
argument to the outer function, so that a MOV instruction
is needed to take arg1 -> reg0. This makes the transformation
easier to read than just making the entire function a NOP.

LewisCrawford commented on llvm/test/CodeGen/X86/fmaxnum.ll:
```diff
@@ -653,5 +653,29 @@ define float @test_maxnum_const_nan(float %x) {
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    retq
+  %r = call nnan float @llvm.maxnum.f32(float %x, float 0xfff0000000000000)
+  ret float %r
```
Making %x an immediate would trigger constant-folding, so would use a different path. Swapping the args would also not work, as it still returns the same value.

I've added a 2nd argument %y to the outer function, and am calling min/max on %y instead now, so that the compiler needs to move arg1 into the return register, rather than just relying on arg0 already being in the right place. Hopefully this makes things easier to read and less implicit.

LewisCrawford committed:
Merge branch 'main' into dagcombine_maximumnum

LewisCrawford committed:
Merge branch 'main' into dagcombine_maximumnum
